### PR TITLE
Builtin cache invalidation system aka make API Platform fast as hell

### PR DESCRIPTION
**Pull request by dunglas**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | n/a
| License       | MIT
| Doc PR        | todo

The usual quote:

> There are only two hard things in Computer Science: cache invalidation and naming things.
>
> -- Phil Karlton

Well, API Platform is an awesome name, so this PR try to solve the other hard thing: caching to make your API as fast as possible.

It introduces a builtin mechanism to always serve API responses from a cache (Varnish and CloudFlare are targeted), and invalid stale data in real time when a resource is updated, deleted or created.
With this mechanism, on my computer and using Docker, large and complex responses are served in **~15ms instead of ~700ms** without cache.

The API Platform serializer has been tweaked to store the list of all resources included in a given response (the root document, but also embedded documents and documents appearing in lists).

The response is marked with all resources it contains in the `Cache-Tags` HTTP header. A md5 hash of all included IRIs is generated to prevent collisions and reduce the header size.
Then, all API's responses are stored with a high expiration time in the proxy cache. On all subsequent (read) requests from a client, the response is served from the proxy, the PHP application is not touched.

When a resource is modified (only changes made to Doctrine entities are supported for now), all responses containing or referencing it are purged from the cache.
A class to purge Varnish is provided in this PR, and a class to purge CloudFlare (enterprise plans only) will be provided later. An interface allows to add support for other cache providers.

Paged collections are also handled: if a resource is added or removed, all collection pages are purged. If a resource is edited, pages (and all other API responses, including those embedding it as a nested document) are purged.

Enabling this new feature **doesn't require to change existing code**. The following config enable the mechanism and mades your API instantly blazing fast:

```yaml
api_platform:
    http_cache:
        enable_tags: true
        varnish_url: 'http://my-varnish-proxy'
        shared_max_age: 3600
```

I've also opened a PR to add Varnish with a compatible setup to the API Platform Docker setup: api-platform/api-platform#238.
Last but not least, this PR introduces some config options to set global cache settings. Example:

```yaml
api_platform:
    http_cache:
        max_age: 60
        shared_max_age: 3600
        vary: ['Content-Type', 'Cookie']
```

Note: for advanced needs, prefer the awesome [FosHttpCache library](foshttpcache.readthedocs.io).

As stated in the quote, cache invalidation **is** a hard thing and this PR probably contains bugs and edge cases. Please test it and report any problem.

TODO:
* [x] Add unit tests

### CODE REVIEW

File: src/HttpCache/EventListener/PurgeListener.php [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+        $this->purger = $purger;
+        $this->iriConverter = $iriConverter;
+    }
+
+    /**
+     * Stores the original IRI of the object to be deleted because it may not be available anymore at purge time (e.g. when using Doctrine).
+     *
+     * @param GetResponseForControllerResultEvent $event
+     */
+    public function onKernelView(GetResponseForControllerResultEvent $event)
```
Comment (dunglas) [inline src/HttpCache/EventListener/PurgeListener.php]: Probably useless because it is already done for Doctrine and not reliable enough for non-Doctrine.

File: src/HttpCache/EventListener/PurgeListener.php [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    {
+        $this->addResources($event->getRequest());
+
+        if ($iris = $this->resourceManager->getResources()) {
+            $this->purger->purge($iris);
+            $this->resourceManager->clear();
+        }
+    }
+
+    private function addResources(Request $request)
```
Comment (dunglas) [inline src/HttpCache/EventListener/PurgeListener.php]: Probably useless because it is already handled for Doctrine and not reliable enough for non-Doctrine data providers.

**Comment by fbourigault** [comment]
This looks amazing, but does providing a solution coupled to specific cache solution as first class citizen a good idea?

Maybe an intermediate solution build using standard Http headers would improve interoperability.

**Comment by bendavies** [comment]
Nice work @dunglas.

This is a very similar strategy to what i've implemented previously.  
A few comments:

1. why not just use FosHttpCache instead of implementing your own tagging/purgers etc?  
2. It does no look like this will work with authenticated apis that don't use cookies (FosHttpCache has provisions for user specific caching)
3. why do you md5 the tags?

File: composer.json [hunk 3]
```diff
@@ -34,6 +34,7 @@
         "doctrine/orm": "^2.5",
         "doctrine/annotations": "^1.2",
         "friendsofsymfony/user-bundle": "^2.0@dev",
+        "guzzlehttp/guzzle": "^6.0",
```
Comment (fbourigault) [inline composer.json]: You should use `php-http/httplug` for better interop.

**Comment by dunglas** [comment]
@fbourigault there is an abstraction layer and and implementation (Varnish). Basically, anyone can add support for any cache provider supporting cache invalidation (it's a matter of implementing an interface). I plan to add support CloudFlare in core too.

There is no standard (HTTP headers) for cache invalidation (only expiration is supported in RFCs) but I choose to use `Cache-Tags` because it's the one used by CloudFlare.

@bendavies regarding FosHttpCache, for 2 reasons:

* my implementation is a bit different (the md5 hash) and tight to the concept of "resources" and "IRIs", not present in FosHttpCache
* the implementation is trivial (and it's possible to bridge it with FosHttpCache if wanted), it will ease our maintenance process to not have a dependency to a 3rd party library we don't maintain (our soft dependency to FosUser is a pain to maintain...).

It can work with any mechanism of authentification (it's why I've introduced a way to configure the `Vary` HTTP header). If the resource varies depending of the `Authorization` header, just set `api_platform.vary` to `['Content-Type', 'Authorization']` (you'll need to tweak the Varnish config too). A key concept of REST is being explicit, a resource identified by an IRI should not vary depending of the current logged in user (in this case, the URL should not be the same).

3/ To avoid collision (if the a resource is tagged with `/foos/,/foos1` and another with only `/foos/2`, sending a BAN request with `/foos` as parameter will ban both, and it's a bug. Using md5 hashes prevent this problem. A CRC32 hash should do the trick too, but md5 is usually faster on modern servers.

File: composer.json (hunk 3 above)
Comment (dunglas) [inline composer.json]: Maybe when it will be a PSR. But for now it has nos advantage over using Guzzle 6 for this use case: it is harder to setup (requires to install an extra Symfony bundle), adds complexity and a performance overhead.
Guzzle 6 is already an abstraction layer, with builtin curl and streams implementations. We only depend of the `ClientInterface`, it's easy to bridge other implementations (but I think it is not worth it).

As we support PHP 7 only, we don't target older libraries such as Guzzle 5.

Comment (fbourigault) [inline composer.json]: I didn't catch the last point. :+1:

**Comment by fbourigault** [comment]
> There is no standard (HTTP headers) for cache invalidation (only expiration is supported in RFCs) but I choose to use Cache-Tags because it's the one used by CloudFlare.

By standard HTTP headers, I mean revalidation. But maybe in such case, caching is not efficient.

**Comment by teohhanhui** [comment]
> There is no standard (HTTP headers) for cache invalidation (only
expiration is supported in RFCs) but I choose to use Cache-Tags because
it's the one used by CloudFlare.

Custom HTTP headers should be prepended by `X-`. CloudFlare is flaunting
the spec (bad).

> If the resource varies depending of the Authorization header, just
set api_platform.vary to ['Content-Type', 'Authorization'] (you'll need to
tweak the Varnish config too).

It's not so simple. An authorization context needs to be provided and
hashed in Varnish. Blindly adding `Vary: Authorization` would kill hit
rates, if your intention is to keep a cache for each user, for example.

> A key concept of REST is being explicit, a resource identified by an IRI
should not vary depending of the current logged in user (in this case, the
URL should not be the same).

False. A resource can vary by any header as declared in the Vary header.

**Comment by teohhanhui** [comment]
> To avoid collision (if the a resource is tagged with /foos/,/foos1 and
another with only /foos/2, sending a BAN request with /foos as parameter
will ban both, and it's a bug. Using md5 hashes prevent this problem. A
CRC32 hash should do the trick too, but md5 is usually faster on modern
servers.

Just use regex. Varnish has good regex support. Using a hash is a nightmare
when trying to debug cache tags. :)

**Comment by dunglas** [comment]
I'll make the header configurable but I'll keep CloudFlare compatibility by default, it's an invaluable feature. By the way this header should not be exposed to the end client.

Regarding Vary, it's exactly what I explain in my post 🙂

Regarding the hit rate, doing more advanced things like hashing cannot be automated on the API Platform side, it requires a custom development and it's easy to implement using the new vary option.
Using FosHTTPCache for such usages is also an option, it's up to the developer.

**Comment by dunglas** [comment]
Regex (at least simple regexes) doesn't fix the issue in my example. /foos will match /foos/*, not only the collection response.

**Comment by teohhanhui** [comment]
/^\/foos$/

**Comment by dunglas** [comment]
But maybe a more complex regex can do the trick, I'll give it a try (I agree that plain tags are better than hashes for debug).

**Comment by dunglas** [comment]
The comma must be handled too.

File: features/http_cache/tags.feature [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    When I add "Content-Type" header equal to "application/ld+json"
+    And I send a "PUT" request to "/relation_embedders/1" with body:
+    """
+    {
+      "paris": "France"
+    }
+    """
+    Then the response status code should be 200
+    And the header "Cache-Tags" should not exist
+    And "/relation_embedders/1,/related_dummies/1,/third_levels/1" IRIs should be purged
```
Comment (jderusse) [inline features/http_cache/tags.feature]: A PUT may change the value of an attribute used to order a responses. You should invalidate collection too

File: features/http_cache/tags.feature [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+  @dropSchema
+  Scenario: Purge item and the related collection on update
+    When I add "Content-Type" header equal to "application/ld+json"
+    And I send a "DELETE" request to "/relation_embedders/1"
+    Then the response status code should be 204
+    And the header "Cache-Tags" should not exist
+    And "/relation_embedders/1,/relation_embedders" IRIs should be purged
```
Comment (jderusse) [inline features/http_cache/tags.feature]: What's about `related_dummies` who was previously linked to this entity? They should be invalidated too

File: features/http_cache/tags.feature [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    {
+      "anotherRelated": {
+        "name": "Related",
+        "thirdLevel": {}
+      }
+    }
+    """
+    Then the response status code should be 201
+    And the header "Cache-Tags" should not exist
+    And "/relation_embedders,/related_dummies,/third_levels,/relation_embedders/1,/related_dummies/1,/third_levels/1" IRIs should be purged
```
Comment (jderusse) [inline features/http_cache/tags.feature]: I don't understand in you code how you make the link between `relation_embedders` and `third_levels`? Who flaged it?

File: features/http_cache/tags.feature [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+
+  Scenario: Tags must be set for items
+    When I send a "GET" request to "/relation_embedders/1"
+    Then the response status code should be 200
+    And the header "Cache-Tags" should be equal to "aa9e2bee5be20590f7dcc520ce2dffca,12a0c94f947a680d68bd6f65e025457d,91774d67418192a057e25dae00345572"
```
Comment (jderusse) [inline features/http_cache/tags.feature]: md5 obfuscate tests... :( 
I wonder if a dedicated context could be more readable: `the cache-tag header should be equals to "/relation_embedders, ...."`

**Comment by bendavies** [comment]
@dunglas 
https://github.com/FriendsOfSymfony/FOSHttpCache/blob/master/src/ProxyClient/Varnish.php#L59-L61
https://regex101.com/r/vqmNLw/2

File: composer.json (hunk 3 above)
Comment (lyrixx) [inline composer.json]: @dunglas IMHO, you should consider using buzz. And It's not a joke:

* Guzzle have so many version, and these versions are not compatible each others. So people that are using a version that is incompatible with guzzle 6 will be blocked.
* Buzz is very simple and works well

File: features/http_cache/tags.feature (hunk 5 above)
Comment (dunglas) [inline features/http_cache/tags.feature]: They will be, because they are marked with `/relation_embedders/1`.

File: features/http_cache/tags.feature (hunk 4 above)
Comment (dunglas) [inline features/http_cache/tags.feature]: Good catch, I add this.

File: features/http_cache/tags.feature (hunk 6 above)
Comment (dunglas) [inline features/http_cache/tags.feature]: It's because of this line: https://github.com/api-platform/core/pull/952/files#diff-0362bfae91549bcd08af0d06bd806b37R14

File: features/http_cache/tags.feature (hunk 7 above)
Comment (dunglas) [inline features/http_cache/tags.feature]: True, I'm working on a fix to remove the hash and use plain IRIs. I'll update the PR when it's ready.

File: src/Bridge/Symfony/Bundle/DependencyInjection/Configuration.php [hunk 8]
```diff
@@ -92,6 +92,23 @@ public function getConfigTreeBuilder()
...
+                        ->integerNode('max_age')->defaultNull()->info('Default value for the response max age.')->end()
+                        ->integerNode('shared_max_age')->defaultNull()->info('Default value for the response shared (proxy) max age.')->end()
+                        ->arrayNode('vary')
+                            ->defaultValue(['Content-Type'])
+                            ->prototype('scalar')->end()
+                            ->info('Default values of the "Vary" HTTP header.')
+                        ->end()
+                        ->booleanNode('public')->defaultNull()->info('To make all responses public by default.')->end()
+                        ->booleanNode('enable_tags')->defaultFalse()->info('Add cache tags to the response.')->end()
+                        ->scalarNode('varnish_url')->defaultNull()->info('URL of the Varnish server to purge using cache tags when a resource is updated.')->end()
```
Comment (jderusse) [inline src/Bridge/Symfony/Bundle/DependencyInjection/Configuration.php]: Don't forget user with several varnish servers

**Comment by dunglas** [comment]
Thank you for the reviews everyone!

* Plain IRIs are now used instead of md5 hashes (and the regex have been updated accordingly)
* Collections are purged even when a PUT is done

File: composer.json (hunk 3 above)
Comment (dunglas) [inline composer.json]: AFAIK it doesn't support async requests... and it has not been updated since 2015.

Comment (fbourigault) [inline composer.json]: When guzzle guys will release the 7.0 version, you will be stuck with the 6.0 version. Http clients are a widely common dependency and so conflicts may appears. But ATM you could go with guzzle 6. Using httplug could be added later.

File: src/Bridge/Doctrine/EventListener/PurgeHttpCacheListener.php [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+     */
+    public function __construct($container)
+    {
+        $this->container = $container;
+    }
+
+    public function onFlush(OnFlushEventArgs $eventArgs)
+    {
+        $iriConverter = $this->container->get('api_platform.iri_converter');
+        $resourceManager = $this->container->get('api_platform.http_cache.resource_manager');
```
Comment (soyuka) [inline src/Bridge/Doctrine/EventListener/PurgeHttpCacheListener.php]: Circular reference is only because of IriConverter no? Can't you inject the ResourceManager properly though?

File: src/HttpCache/VarnishPurger.php [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+    /**
+     * {@inheritdoc}
+     */
+    public function purge(array $iris)
+    {
+        if (!$iris) {
+            return;
+        }
+
+        $parts = array_map(function ($iri) {
```
Comment (soyuka) [inline src/HttpCache/VarnishPurger.php]: A small comment to explain what goes below would be great for future readers!

File: src/Bridge/Symfony/Bundle/Resources/config/http_cache_tags.xml [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+
+            <tag name="kernel.event_listener" event="kernel.response" method="onKernelResponse" />
+        </service>
+
+        <!-- The onKernelView method must be executed after the validation (32) but before the write (64) -->
+        <service id="api_platform.http_cache.listener.response.purge" class="ApiPlatform\Core\HttpCache\EventListener\PurgeListener">
+            <argument type="service" id="api_platform.http_cache.resource_manager" />
+            <argument type="service" id="api_platform.http_cache.purger" />
+
+            <tag name="kernel.event_listener" event="kernel.terminate" method="onKernelTerminate" />
```
Comment (jderusse) [inline src/Bridge/Symfony/Bundle/Resources/config/http_cache_tags.xml]: I suggest to move the purge as close as possible to the transaction (doctrine postFlush IMO)

I understand your point of view to return a response to the user as soon as possible, but this introduce cross concurrency bugs.

Example:
- list all blogs (cache mis)
- post a blog
- list all blog (cache hit when called before the kernel terminate even)

I already have this issue in the my CI (agreed, this is a particular case with a lot of stress tests). but the probability of cross concurrency is higher when:
- you have several varnishes to purge
- you have other kernel.terminate events with an higher priority (sending an email for instance)

File: src/HttpCache/EventListener/ConfigureListener.php [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+ * @author Kévin Dunglas <dunglas@gmail.com>
+ */
+final class ConfigureListener
+{
+    private $maxAge;
+    private $sharedMaxage;
+    private $vary;
+    private $public;
+
+    public function __construct(int $maxAge = null, int $sharedMaxAge = null, array $vary, bool $public = null)
```
Comment (jderusse) [inline src/HttpCache/EventListener/ConfigureListener.php]: I suggest to add a LastModified header too initialized to the default value `now()`.
As a result, Varnish will handled it for free (by default), and returns 304 code stuff.

This small change should not interfer in your App and how you invalid cache. But will save bandwidth between the user and varnish.

File: src/Bridge/Symfony/Bundle/DependencyInjection/ApiPlatformExtension.php [hunk 13]
```diff
@@ -288,6 +296,41 @@ private function registerDoctrineExtensionConfiguration(ContainerBuilder $contai
         }
     }
 
+    private function registerHttpCache(ContainerBuilder $container, array $config, XmlFileLoader $loader)
+    {
+        $loader->load('http_cache.xml');
+
+        if (true !== $config['http_cache']['enable_tags']) {
```
Comment (jderusse) [inline src/Bridge/Symfony/Bundle/DependencyInjection/ApiPlatformExtension.php]: When `enable_tags` is true and `maxAge` > 0. We should warn the developer that the browser will cache responses and remove the benefits of cache invalidation.

File: src/Bridge/Symfony/Bundle/DependencyInjection/Configuration.php [hunk 14]
```diff
@@ -92,6 +92,27 @@ public function getConfigTreeBuilder()
...
+                    ->children()
+                        ->integerNode('max_age')->defaultNull()->info('Default value for the response max age.')->end()
+                        ->integerNode('shared_max_age')->defaultNull()->info('Default value for the response shared (proxy) max age.')->end()
+                        ->arrayNode('vary')
+                            ->defaultValue(['Content-Type'])
+                            ->prototype('scalar')->end()
+                            ->info('Default values of the "Vary" HTTP header.')
+                        ->end()
+                        ->booleanNode('public')->defaultNull()->info('To make all responses public by default.')->end()
+                        ->booleanNode('enable_tags')->defaultFalse()->info('Add cache tags to the response.')->end()
```
Comment (jderusse) [inline src/Bridge/Symfony/Bundle/DependencyInjection/Configuration.php]: it's not only about adding tags in response, it's also about cache invalidation.

File: src/HttpCache/VarnishPurger.php [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+            return sprintf('(^|\,)%s($|\,)', preg_quote($iri));
```
Comment (jderusse) [inline src/HttpCache/VarnishPurger.php]: You may store the tag with a `,` at the beginning and ending to simplify this RegExp and optimizeinvalidation (especially when you invalidate several tags at the time).
Regexp would became `,%s,`

File: src/Bridge/Doctrine/EventListener/PurgeHttpCacheListener.php [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+        $uow = $eventArgs->getEntityManager()->getUnitOfWork();
+
+        foreach ($uow->getScheduledEntityInsertions() as $entity) {
+            $this->purge($iriConverter, $resourceManager, $entity, false);
```
Comment (jderusse) [inline src/Bridge/Doctrine/EventListener/PurgeHttpCacheListener.php]: are you sure to be able to retrieves an auto-increment id in the `onFlush` eventListener? I thought, this ID was injected *after* the insert operation

File: src/HttpCache/EventListener/ConfigureListener.php (hunk 12 above)
Comment (teohhanhui) [inline src/HttpCache/EventListener/ConfigureListener.php]: Varnish already returns 304 response properly. We should not send bogus `Last-Modified` headers.

File: src/Bridge/Symfony/Bundle/Resources/config/http_cache_tags.xml (hunk 11 above)
Comment (teohhanhui) [inline src/Bridge/Symfony/Bundle/Resources/config/http_cache_tags.xml]: ^ That does not actually address the race condition, but only trying to avoid it.

Comment (jderusse) [inline src/Bridge/Symfony/Bundle/Resources/config/http_cache_tags.xml]: indeed, it fixes one case.

BTW (not related to the RC): A small improvment for the user would be to start async request on the `postFlush` event, and wait on the `kernelFinish` event.

File: src/HttpCache/EventListener/ConfigureListener.php (hunk 12 above)
Comment (jderusse) [inline src/HttpCache/EventListener/ConfigureListener.php]: I thought this listener was used only with "tag" thing. I apologize, agreed, this header don't make sens here.

But I think, you missed my point, the idea, when using tagged response, is to add a validation header (could be etag or last-modified, don't care). To save bandwidth between users and varnish when cache is hit.

Comment (dunglas) [inline src/HttpCache/EventListener/ConfigureListener.php]: IMO we should provide a good default config for this (in the standard edition) but let the user configure it as he wants.

Comment (teohhanhui) [inline src/HttpCache/EventListener/ConfigureListener.php]: @jderusse:
> the idea, when using tagged response, is to add a validation header (could be etag or last-modified, don't care). To save bandwidth between users and varnish when cache is hit.

That's incorrect. Varnish already sends 304 response with proper `E-Tag` in that case. There is no need for the backend to send an `E-Tag` or `Last-Modified` header unless the backend intends to return 304 on `vcl_backend_fetch`.

File: composer.json (hunk 3 above)
Comment (teohhanhui) [inline composer.json]: I have a very strong (I think) argument for using HTTPlug: you do not get into hell like FOSHttpCache did (and still is) with their hard coupling to a specific Guzzle version.

File: src/HttpCache/EventListener/ConfigureListener.php (hunk 12 above)
Comment (jderusse) [inline src/HttpCache/EventListener/ConfigureListener.php]: Of course varnish deals with validation headers and return 304 responses. I'd never said to handle the validation in the app. BUT if you don't send validation headers in the response, clients will never send validations headers back in the next request (if-none-match and if-modified-since) and varnish will never have the possibility to use it. Here is my point.
It's free, it don't change the behavior of your application (if you don't deal invalidation on app side) and it save bandwidth.

Comment (teohhanhui) [inline src/HttpCache/EventListener/ConfigureListener.php]: > BUT if you don't send validation headers in the response, clients will never send validations headers back in the next request (if-none-match and if-modified-since) and varnish will never have the possibility to use it.

Varnish does it. :smile:

Comment (jderusse) [inline src/HttpCache/EventListener/ConfigureListener.php]: Let explain with a schema (my last try)...
Here is the actual implementation
now

My objective is to transform the last response in step **6** into a 304.

* To do it, varnish need a `if-none-match` header in step **5**
* To do it, client need a `etag` header in step **4**
* To do it, varnish needs nothing, it just forward the response from the app from step **3**
* To do it, app must send this header <= this is my point

And Here is the target sequence
then



Your last comment suggest that varnish add this header by it self, with something like that
other

Is it what you mean? Because I doubt varnish add such header by itself (unless you implement this behavior in a VCL which is fine to me too)

Comment (teohhanhui) [inline src/HttpCache/EventListener/ConfigureListener.php]: Varnish does it by default. Try it. At least it has been doing so since 4.0 :smile:

File: src/Bridge/Doctrine/EventListener/PurgeHttpCacheListener.php (hunk 16 above)
Comment (jderusse) [inline src/Bridge/Doctrine/EventListener/PurgeHttpCacheListener.php]: nevermind, if missed the option `false` which means: don't purge the item itself :)

**Comment by jderusse** [comment]
beware, this PR sends cache headers on error 500 too
```
curl test2_nginx_1.docker/foos -I
HTTP/1.1 500 Internal Server Error
Content-Type: application/ld+json; charset=utf-8
Cache-Control: public, s-maxage=3600
Link: <http://test2_nginx_1.docker/docs.jsonld>; rel="http://www.w3.org/ns/hydra/core#apiDocumentation"
Vary: Content-Type
Cache-Tags: /foos
Date: Wed, 22 Mar 2017 07:20:44 GMT
```

File: src/HttpCache/EventListener/ConfigureListener.php (hunk 12 above)
Comment (jderusse) [inline src/HttpCache/EventListener/ConfigureListener.php]: I bootstraped an empty api-plateform from this PR https://github.com/api-platform/api-platform/pull/238

```
curl test2_varnish_1.docker/foos -i
HTTP/1.1 200 OK
Server: nginx/1.11.10
Content-Type: application/ld+json; charset=utf-8
X-Powered-By: PHP/7.1.3
Vary: Content-Type
X-Content-Type-Options: nosniff
X-Frame-Options: deny
Cache-Control: public, s-maxage=3600
Link: <http://test2_varnish_1.docker/docs.jsonld>; rel="http://www.w3.org/ns/hydra/core#apiDocumentation"
Cache-Tags: /foos
Date: Wed, 22 Mar 2017 07:23:50 GMT
X-Varnish: 26 32789
Age: 3
Via: 1.1 varnish-v4
Accept-Ranges: bytes
Content-Length: 111
Connection: keep-alive

{"@context":"\/contexts\/Foo","@id":"\/foos","@type":"hydra:Collection","hydra:member":[],"hydra:totalItems":0}
```

varnish don't add validation header :confused:

File: src/HttpCache/EventListener/AddTagsListener.php [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+     *
+     * @param FilterResponseEvent $event
+     */
+    public function onKernelResponse(FilterResponseEvent $event)
+    {
+        $request = $event->getRequest();
+        if (
+            !$request->isMethodCacheable()
+            || !$attributes = RequestAttributesExtractor::extractAttributes($request)
+        ) {
```
Comment (teohhanhui) [inline src/HttpCache/EventListener/AddTagsListener.php]: To address @jderusse's [comment](https://github.com/api-platform/core/pull/952#issuecomment-288318323), we'd add another condition here:

```php
!in_array($response->getStatusCode(), [
    Response::HTTP_OK,
    Response::HTTP_NO_CONTENT,
])
```

File: src/HttpCache/EventListener/ConfigureListener.php (hunk 12 above)
Comment (teohhanhui) [inline src/HttpCache/EventListener/ConfigureListener.php]: @jderusse You're right. I overlooked that FOSHttpCacheBundle was the one adding it. My apologies...

I think we should use the same approach. Just use the MD5 hash of the `Response` content as `ETag`. MD5 is cheap, so no reason to use a meaningless `Last-Modified`. :smile:

**Comment by dunglas** [comment]
I'll refactor this PR to get rid of the stateful service.

**Comment by dunglas** [comment]
About

> Custom HTTP headers should be prepended by `X-`. CloudFlare is flaunting
the spec (bad).

This convention has been deprecated in RFC 6648. See https://specs.openstack.org/openstack/api-wg/guidelines/headers.html. I'll remove all `X-` prefix from this PR.

File: src/Bridge/Doctrine/EventListener/PurgeHttpCacheListener.php [hunk 18]
```diff
@@ -0,0 +1,81 @@
...
+        $this->container = $container;
+    }
+
+    public function onFlush(OnFlushEventArgs $eventArgs)
+    {
+        if (!$request = $this->container->get('request_stack')->getCurrentRequest()) {
+            return;
+        }
+
+        $resources = $request->attributes->get('_resources', []);
```
Comment (jderusse) [inline src/Bridge/Doctrine/EventListener/PurgeHttpCacheListener.php]: This tight couple the purge system with http requests context which is used like a databag.

Comment (dunglas) [inline src/Bridge/Doctrine/EventListener/PurgeHttpCacheListener.php]: Yes I'll try to find a better solution but I'm not sure there is one. Anyway its called `purge HTTP cache`, it's not a big deal to couple it with the HTTP request.

File: src/HttpCache/EventListener/ConfigureListener.php (hunk 12 above)
Comment (dunglas) [inline src/HttpCache/EventListener/ConfigureListener.php]: Doing a md5 of the response is not ok (the response contain the date). We may do a md5 of the response's body.

File: src/Bridge/Symfony/Bundle/DependencyInjection/ApiPlatformExtension.php (hunk 13 above)
Comment (dunglas) [inline src/Bridge/Symfony/Bundle/DependencyInjection/ApiPlatformExtension.php]: It can be intended (I often do that). It allows to always serve fresh data for the first request but reduce the server load/bandwidth using expiration. It's often not a problem when the TTL is low. I would let the user do what he wants without warning.

File: src/HttpCache/VarnishPurger.php (hunk 15 above)
Comment (dunglas) [inline src/HttpCache/VarnishPurger.php]: I'm not a fond of this, the header become ugly. It may have other use cases for this headers (e.g. : building a smart client-side cache).

File: src/HttpCache/EventListener/AddTagsListener.php (hunk 17 above)
Comment (dunglas) [inline src/HttpCache/EventListener/AddTagsListener.php]: done

**Comment by dunglas** [comment]
All comments handled. Can you make a last review?

File: src/Hal/Serializer/ItemNormalizer.php [hunk 19]
```diff
@@ -46,13 +46,16 @@ public function normalize($object, $format = null, array $context = [])
...
+        $context['api_normalize'] = true;
 
         $rawData = parent::normalize($object, $format, $context);
         if (!is_array($rawData)) {
             return $rawData;
         }
 
-        $data = ['_links' => ['self' => ['href' => $this->iriConverter->getIriFromItem($object)]]];
+        $data = ['_links' => ['self' => ['href' => $context['iri']]]];
+        $context['debug'] = true;
```
Comment (sroze) [inline src/Hal/Serializer/ItemNormalizer.php]: Is there any reason for this property in the context?

File: features/http_cache/tags.feature [hunk 20]
```diff
@@ -0,0 +1,64 @@
+Feature: Cache invalidation trough HTTP Cache tags
```
Comment (soyuka) [inline features/http_cache/tags.feature]: through?

File: src/HttpCache/EventListener/AddTagsListener.php [hunk 21]
```diff
@@ -0,0 +1,71 @@
...
+    public function onKernelResponse(FilterResponseEvent $event)
+    {
+        $request = $event->getRequest();
+        $response = $event->getResponse();
+
+        if (
+            !$request->isMethodCacheable()
+            || !$response->isCacheable()
+            || (!$attributes = RequestAttributesExtractor::extractAttributes($request))
+            || !$resources = $request->attributes->get('_resources')
```
Comment (soyuka) [inline src/HttpCache/EventListener/AddTagsListener.php]: no parenthesis here, but the line above yes?

File: tests/Serializer/ItemNormalizerTest.php [hunk 22]
```diff
@@ -94,7 +95,7 @@ public function testNormalize()
         );
         $normalizer->setSerializer($serializerProphecy->reveal());
 
-        $this->assertEquals(['name' => 'hello'], $normalizer->normalize($dummy));
+        $this->assertEquals(['name' => 'hello'], $normalizer->normalize($dummy, null, ['resources' => []]));
```
Comment (soyuka) [inline tests/Serializer/ItemNormalizerTest.php]: Shouldn't we avoid changing those tests? It feels like the test won't pass without 2 more arguments (eg, `null, ['resources' => []]`). Is it the case / Isn't this breaking things somehow?

File: src/Hal/Serializer/ItemNormalizer.php (hunk 19 above)
Comment (dunglas) [inline src/Hal/Serializer/ItemNormalizer.php]: Good catch, it has nothing to do here!

File: src/HttpCache/EventListener/AddTagsListener.php (hunk 21 above)
Comment (dunglas) [inline src/HttpCache/EventListener/AddTagsListener.php]: The parenthesis on the previous line are mandatory because of operator priorities.

File: tests/Serializer/ItemNormalizerTest.php (hunk 22 above)
Comment (dunglas) [inline tests/Serializer/ItemNormalizerTest.php]: It passes without the arguments and without change, but I've modified it to test the new behavior.

**Comment by Simperfit** [comment]
I'm going to test this with a project, ill tell if there are some bugs

File: tests/Serializer/ItemNormalizerTest.php (hunk 22 above)
Comment (soyuka) [inline tests/Serializer/ItemNormalizerTest.php]: Okay as long as there is a test that doesn't adds up those arguments.

**Comment by patkar** [comment]
May you use http://httplug.io/ instead of guzzle directly?

**Comment by teohhanhui** [comment]
> May you use http://httplug.io/ instead of guzzle directly?

It has been raised before, and I'd like to once again echo this.

**Comment by dunglas** [comment]
HTTPPlug introduces a lot of complexity (including an extra bundle to configure... until we get Flex) for no gain here. I'm still 👎 for now.
I understand the problem with Guzzle 3, but API Platform is younger than Guzzle 3, every bundle I use are compatible with Guzzle 6, and AFAIK, there is no projet to develop Guzzle 7 soon.

**Comment by dunglas** [comment]
And by the way, Guzzle is a soft dependency here. So someone wanting to use another client can do it, he just have to implement himself the `PurgerInterface`, not a big deal.

**Comment by theofidry** [comment]
As long as Guzzle is a soft dependency it's ok
